pycket/pycket
# Pycket with Racket 8

#### Description

This brings Pycket up to speed with Racket.

* Using the most recent expander, fasl, and regexp linklets.
* Linklet form is slightly changed in Racket 8 (or 7), this adapts to that (but doesn't break compatibility with older version.
* Adds some new data structures such as `W_EqualAlwaysHashTable`, and a new equality `equal-always?`.
* The tests are passing with both `old` and `new` Pyckets.
* Compiling racket modules and loading from zo files are working, we're able to load `racket/base` fully from compiled code.

Probably some more stuff I don't recall at the moment.

#### QA Steps

Compile:

```
make pycket-c-linklets
```

The bootstrapping linklets are already a part of the repo, but you can also regenerate them if you like to make sure everything's working:

```
make expander && make fasl && make regexp
```

Setup Racket and `PLTHOME` variable and run the following to load a repl with `racket/base`:

```
./pycket-c-linklets --verbose 3
```

You may run `make compile-racket-modules` to compile and generate zo files for racket modules. After that loading `racket/base` should be much faster.

```
make compile-racket-modules && ./pycket-c-linklets --verbose 3
```

This will be adding a Github Action for build/CI/CD, but you can also run the tests yourself to QA this.

```
make test-old && make test-new
```

#### Notes & Discussions

I still see some issues in loading the `typed/racket`, but we're getting there.

Some next steps include loading the `io` linklet, tidy up and rebase the `stackful` branch, and more.

**I'll merge this and set up the Github Actions in a separate PR.**
